Add Security document (#766)
canonical/synapse-operator: docs/explanation/security.md

  - [federation_domain_whitelist](https://charmhub.io/synapse/configurations#federation_domain_whitelist): Comma-separated list of allowed federation domains.
  - [invite_checker_blocklist_allowlist_url](https://charmhub.io/synapse/configurations#invite_checker_blocklist_allowlist_url): Configure allowed and blocked lists for invite sources.
  - [ip_range_whitelist](https://charmhub.io/synapse/configurations#ip_range_whitelist): Comma-separated list of allowed IP address ranges (CIDR).
  - [limit_remote_rooms_complexity](https://charmhub.io/synapse/configurations#limit_remote_rooms_complexity): Limits the complexity of remote rooms; if a room exceeds this complexity, users will be prevented from joining.

### Security vulnerabilities

Running Synapse with one or more weakness that can be exploited by attackers.

#### Best practices

  - Keep the Juju and the charm updated. See more about Juju updates in the [documentation](https://documentation.ubuntu.com/juju/latest/explanation/juju-security/index.html#regular-updates-and-patches).

### Unencrypted traffic

> When HTTPS is not enabled, data exchanged between Synapse and its clients — including authentication tokens, registration secrets, and personal information — is transmitted in plain text. This makes it vulnerable to interception, tampering, and impersonation through [man-in-the-middle (MitM)](https://www.sciencedirect.com/topics/computer-science/man-in-the-middle-attack) attacks.

Author: The warning will be ignored and we will keep the term man-in-the-middle since is a common reference.

> When HTTPS is not enabled, data exchanged between Synapse and its clients — including authentication tokens, registration secrets, and personal information — is transmitted in plain text. This makes it vulnerable to interception, tampering, and impersonation through [man-in-the-middle (MitM)](https://www.sciencedirect.com/topics/computer-science/man-in-the-middle-attack) attacks.

Author: The warning will be ignored and we will keep the term man-in-the-middle since is a common reference.

#### Best practices

- Always enable HTTPS:

>   Configure Synapse to use HTTPS for all clients communication. The Synapse charm supports ingress integration, allowing HTTPS to be enabled when integrating with charms such as [nginx-ingress-integrator](https://charmhub.io/nginx-ingress-integrator) and [traefik](https://charmhub.io/traefik-k8s).

Reviewer: ```suggestion
  Configure Synapse to use HTTPS for all clients communication. The Synapse charm supports ingress integration, allowing HTTPS to be enabled when integrating with charms such as [nginx-ingress-integrator](https://charmhub.io/nginx-ingress-integrator) and [traefik-k8s](https://charmhub.io/traefik-k8s).
```
Nitpick

# Security

This document outlines common risks and possible best practices specifically for the Synapse charm. It

> focuses on configurations and protections available through the charm itself.

Reviewer: In this document, do you think it's worth adding any information about built-in security features that the Synapse charm has (for example authentication, authorization, or cryptography)?

Author: Related specifically to the charm, I don’t see how these would fit here:

- **Authentication:** With MAS in place, there’s already [this document](https://github.com/canonical/synapse-operator/blob/2/main/docs/how-to/migrate-to-mas.md#configure-oidc) explaining OIDC, including links to the upstream documentation.
- **Authorization:** There’s nothing specific to the charm for this — it’s only relevant when working directly with Synapse.
- **Cryptography:** This is used during the integration between Synapse and the IRC bridge, but there’s no user-facing action involved. It feels more appropriate for a charm library explanation or for the event flow architecture we discussed.

However, this actually reminded me of something that could be added here:  
A recommendation to place **Traefik or Nginx** in front of Synapse, since a valid TLS certificate is required.

wdyt

Reviewer: Thank you for the additional context! 

For authentication and authorization, I agree with you that we don't need to take any further action in the documentation. 

For cryptography, I think that this could be documented better and I'm fine with covering it in a separate explanation document (even if it's short). I'd prefer for this document to be included in this PR. Let me know if you have any concerns or questions there.

I think it's a great idea to include the Traefik or Nginx recommendation in this document. Do you think it's a best practice for a specific risk, or should it go under a separate header like `## Other recommendations`?

Author: I added the new doc now and about the https, I added it as a new risk/best practice. And the "man-in-the-middle" marked as insensitive is new to me. I'm afraid that replacing it with the suggestions would be confusing, all references use this term.

Reviewer: Hey, thanks so much for adding the new document and additional information! 

For now we can ignore since Vale checks are not mandatory or blocking, but if you feel strongly that we shouldn't replace "man-in-the-middle" with one of Vale's suggestions, then we can add an ignore flag to the document that will allow Vale to skip over the term and not flag as an error. We can tackle that in the next cycle :)

Author: I  would prefer keeping the term "man-in-the-middle" because changing it might be really confusing since is a common reference in technical/security documents.

For details regarding upstream Synapse configuration and broader security considerations, please
refer to the [official Synapse documentation](https://element-hq.github.io/synapse/latest/welcome_and_overview.html).

## Risks

The following items include descriptions of the risks, their corresponding best practices for mitigation, as well as links to related documentation and configuration guidelines.

### Loss of data

Synapse configuration files or the database might be destroyed, corrupted, or become inaccessible.

#### Best practices

> - Set up regular backups

Reviewer: ```suggestion
- Set up regular backups:
```

  Follow the [charm documentation](https://charmhub.io/synapse/docs/how-to-backup-and-restore) for guidance on creating backups.

  Synapse uses a file to store its [signing key](https://element-hq.github.io/synapse/latest/usage/administration/backups.html#server-signing-key). If this file is lost, Synapse will generate
  a new one on startup. However, events signed with the old key will no longer be considered valid

>   by other homeservers. So, it's critical having a backup of the charm.

Reviewer: ```suggestion
  by other homeservers, so it's critical to have a backup of the charm.
```

> - Avoid manual media deletion

Reviewer: ```suggestion
- Avoid manual media deletion:
```

  Manually deleting media files will cause Synapse to lose track of them. Always use Synapse's API
  for [media removal](https://element-hq.github.io/synapse/latest/admin_api/media_admin_api.html#purge-remote-media-api) to ensure consistency.

### Data breach

Sensitive data, such as backups containing the signing key or moderation tool access tokens, could
be accessed or stolen by unauthorized individuals.

#### Best practices

> - Secure your backup passphrase

Reviewer: ```suggestion
- Secure your backup passphrase:
```

  Store your [backup_passphrase](https://charmhub.io/synapse/configurations#backup_passphrase) securely and do not share it.

> - Protect your moderation access token secret

Reviewer: ```suggestion
- Protect your moderation access token secret:
```

>   The [moderation_access_token_secret_id](https://charmhub.io/synapse/configurations#moderation_access_token_secret_id) Juju secret contains a token used by Moderation tool. This token

Reviewer: ```suggestion
  The [moderation_access_token_secret_id](https://charmhub.io/synapse/configurations#moderation_access_token_secret_id) Juju secret contains a token used by the Moderation tool. This token
```
Is there a particular reason why "Moderation" is capitalized here?

Author: I thought about "Draupnir" while writing then I remembered to not use a specific technology name, so I renamed it to "Moderation" 😅
My mistake

  grants high power-level access to multiple rooms. Treat it as sensitive and avoid sharing it.

### Unexpected downtime

Synapse may become unavailable, preventing users from accessing the service.

#### Best practices

> - Keep clients updated

Reviewer: ```suggestion
- Keep clients updated:
```

  Synapse performance can be degraded by certain client behaviors, as noted in [Element issue #27867](https://github.com/vector-im/element-web/issues/27867). For example, some clients continuously query `/keys/query` for inactive users, leading to
  performance issues. Keeping clients updated helps avoid known performance problems.

> - Ensure sufficient resources

Reviewer: ```suggestion
- Ensure sufficient resources:
```

  - Plan for adequate storage and database capacity (e.g., media storage and PostgreSQL).
  If the directory used by Synapse becomes full, the service will stop.
  - Consider integrating Synapse with S3 and enabling [enable_media_sync_cleanup](https://charmhub.io/synapse/configurations#enable_media_sync_cleanup) to manage media
  storage more efficiently.

### Denial-of-Service (DoS) attacks

A denial-of-service attack could overwhelm Synapse with traffic, preventing legitimate users from accessing the service.

#### Best practices

> - Rate-limit remote room joins

Reviewer: ```suggestion
- Rate-limit remote room joins:
```

  The Synapse charm provides two configuration options:
  - [rc_joins_remote_burst_count](https://charmhub.io/synapse/configurations#rc_joins_remote_burst_count)
  - [rc_joins_remote_per_second](https://charmhub.io/synapse/configurations#rc_joins_remote_per_second)

  These configurations control the rate at which users can join remote rooms. Refer to the [official documentation](https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#rc_joins) for details on how to configure these values.

> - Restrict invites and federation sources

Reviewer: ```suggestion
- Restrict invites and federation sources:
```
